File: scripts/gh_admin_mod/features/blocklist.py
from pathlib import Path

from gh_admin_mod.env import get_env, normalize_user, parse_bool
from gh_admin_mod.logging import fail
from gh_admin_mod.models import FeatureResult, ModerationContext
from gh_admin_mod.templates import render_template


def _load_blocked_users(blocked_users_file: str) -> set[str]:
    workspace = get_env("GITHUB_WORKSPACE")
    if not workspace:
        fail("GITHUB_WORKSPACE is not available.")

    blocklist_path = Path(workspace, blocked_users_file).resolve()
    if not blocklist_path.exists():
        fail(
            f"Blocked users file not found at {blocklist_path}. "
            "Make sure the repository is checked out before this action runs."
        )

    blocked_users: set[str] = set()
    with open(blocklist_path, "r", encoding="utf-8") as handle:
        for raw_line in handle:
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            blocked_users.add(normalize_user(line))

    return blocked_users


def evaluate(context: ModerationContext) -> FeatureResult:
    if context.is_pull_request and not parse_bool(get_env("INPUT_BLOCK_PRS", "true")):
        return FeatureResult(matched=False)

    if not context.is_pull_request and not parse_bool(get_env("INPUT_BLOCK_ISSUES", "true")):
        return FeatureResult(matched=False)

    blocked_users = _load_blocked_users(get_env("INPUT_BLOCKED_USERS_FILE", "blockedUser.md"))
    normalized_author = normalize_user(context.author)

    if normalized_author not in blocked_users:
        return FeatureResult(matched=False)

    result = FeatureResult(
        matched=True,
        feature="blocklist",
        reason=f"@{context.author} is listed in the blocked users file.",
        blocked_user=context.author,
    )
    result.comment_message = render_template(get_env("INPUT_COMMENT_MESSAGE", ""), context, result)
    return result

File: src/features/blocklist.ts
import * as fs from 'fs';
import * as path from 'path';
import * as core from '@actions/core';
import { getEnv, normalizeUser, parseBool } from '../env.js';
import { FeatureResult, ModerationContext } from '../models.js';
import { renderTemplate } from '../templates.js';

function loadBlockedUsers(blockedUsersFile: string): Set<string> {
  const workspace = getEnv('GITHUB_WORKSPACE');
  if (!workspace) {
    throw new Error('GITHUB_WORKSPACE is not available.');
  }

  const blocklistPath = path.resolve(workspace, blockedUsersFile);
  if (!fs.existsSync(blocklistPath)) {
    throw new Error(
      `Blocked users file not found at ${blocklistPath}. ` +
        'Make sure the repository is checked out before this action runs.'
    );
  }

  const blockedUsers = new Set<string>();
  const content = fs.readFileSync(blocklistPath, 'utf8');
  for (const rawLine of content.split('\n')) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#')) {
      continue;
    }
    blockedUsers.add(normalizeUser(line));
  }

  return blockedUsers;
}

export function evaluate(context: ModerationContext): FeatureResult {
  if (context.is_pull_request && !parseBool(core.getInput('block-prs') || 'true')) {
    return {
      matched: false,
      feature: '',
      reason: '',
      labels: [],
      lock_conversation: false,
      metadata: {},
    };
  }

  if (!context.is_pull_request && !parseBool(core.getInput('block-issues') || 'true')) {
    return {
      matched: false,
      feature: '',
      reason: '',
      labels: [],
      lock_conversation: false,
      metadata: {},
    };
  }

  const blockedUsers = loadBlockedUsers(core.getInput('blocked-users-file') || 'blockedUser.md');
  const normalizedAuthor = normalizeUser(context.author);

  if (!blockedUsers.has(normalizedAuthor)) {
    return {
      matched: false,
      feature: '',
      reason: '',
      labels: [],
      lock_conversation: false,
      metadata: {},
    };
  }

  const result: FeatureResult = {
    matched: true,
    feature: 'blocklist',
    reason: `@${context.author} is listed in the blocked users file.`,
    blocked_user: context.author,
    labels: [],
    lock_conversation: false,
    metadata: {},
  };

  result.comment_message = renderTemplate(core.getInput('comment-message') || '', context, result);

  return result;
}
